File: app/src/main/java/org/mesibo/messenger/MesiboFileTransferHelper.java
/** Copyright (c) 2019 Mesibo
 * https://mesibo.com
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the terms and condition mentioned on https://mesibo.com
 * as well as following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list
 * of conditions, the following disclaimer and links to documentation and source code
 * repository.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or other
 * materials provided with the distribution.
 *
 * Neither the name of Mesibo nor the names of its contributors may be used to endorse
 * or promote products derived from this software without specific prior written
 * permission.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
 * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Documentation
 * https://mesibo.com/documentation/
 *
 * Source Code Repository
 * https://github.com/mesibo/messenger-app-android
 *
 */

package org.mesibo.messenger;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;

import com.mesibo.api.Mesibo;
import com.mesibo.api.MesiboFileTransfer;
import com.mesibo.api.MesiboHttp;

import org.json.JSONObject;
import java.util.LinkedHashMap;
import java.util.Map;

// Refer to http://mesibo.com/documentation/get-started/file-transfer/#how-to-send-and-receive-files
public class MesiboFileTransferHelper implements Mesibo.FileTransferHandler {

    private static MesiboHttp.Queue mQueue = new MesiboHttp.Queue(4, 0);
    public static class MesiboUrl {
        public String op;
        public String file;
        public String result;
        public String xxx;

        MesiboUrl() {
            result = null;
            op = null;
            xxx = null;
            file = null;
        }
    }

    MesiboFileTransferHelper() {
        	//Mesibo.addListener(this);
    }
    
    public boolean Mesibo_onStartUpload(final MesiboFileTransfer file) {

        // we don't need to check origin the way we do in download
        if(mUploadCounter >= 5 && file.priority == 0)
            return false;

        final long mid = file.mid;

        Bundle b = new Bundle();
        b.putString("op", "upload");
        b.putString("token", SampleAPI.getToken());
        b.putLong("mid", mid);
        b.putInt("profile", 0);

        updateUploadCounter(1);
        MesiboHttp http = new MesiboHttp();

        http.url = SampleAPI.getUploadUrl();
        http.postBundle = b;
        http.uploadFile = file.getPath();
        http.uploadFileField = "photo";
        http.other = file;
        file.setFileTransferContext(http);

        http.listener = new MesiboHttp.Listener() {
            @Override
            public boolean Mesibo_onHttpProgress(MesiboHttp config, int state, int percent) {
                MesiboFileTransfer f = (MesiboFileTransfer)config.other;

                if(100 == percent && MesiboHttp.STATE_DOWNLOAD == state) {
                    String response = config.getDataString();
                    String url = null;
                    boolean result = false;

                    try {
                        JSONObject jo = new JSONObject(response);
                        if(null != jo) {
                            url = jo.getString("url");
                            result = jo.getBoolean("result");
                        }

                    } catch (Exception e) {}

                    f.setResult(result, url);
                    return true;
                }

                if(percent < 0) {
                    f.setResult(false, null);
                    return true;
                }

                f.setProgress(percent);
                return true;
            }
        };

        if(null != mQueue)
            mQueue.queue(http);

        else if(http.execute()) {

        }

        return true;

    }

    public boolean Mesibo_onStartDownload(final MesiboFileTransfer file) {

        //TBD, check file type and size to decide automatic download
        if(!SampleAPI.getMediaAutoDownload() && Mesibo.getNetworkConnectivity() != Mesibo.CONNECTIVITY_WIFI && 0 == file.priority)
            return false;

        // only realtime messages to be downloaded in automatic mode.
        if(Mesibo.ORIGIN_REALTIME != file.origin && Mesibo.ORIGIN_DBPENDING != file.origin && file.priority == 0)
            return false;


        updateDownloadCounter(1);

        final long mid = file.mid;

        String url = file.getUrl();
        if(TextUtils.isEmpty(url))
            return false;

        if(!url.toLowerCase().startsWith("http://") && !url.toLowerCase().startsWith("https://")) {
            url = SampleAPI.getDownloadUrl() + url;
        }

        MesiboHttp http = new MesiboHttp();

        http.url = url;
        http.downloadFile = file.getPath();
        http.resume = true;
        http.maxRetries = 10;
        http.other = file;
        file.setFileTransferContext(http);

        http.listener = new MesiboHttp.Listener() {
            @Override
            public boolean Mesibo_onHttpProgress(MesiboHttp config, int state, int percent) {
                MesiboFileTransfer f = (MesiboFileTransfer)config.other;

                if(100 == percent && MesiboHttp.STATE_DOWNLOAD == state) {
                    f.setResult(true, null);
                    return true;
                }

                if(percent < 0) {
                    f.setResult(false, null);
                    return true;
                }

                f.setProgress(percent);
                return true;
            }
        };

        if(null != mQueue)
            mQueue.queue(http);
        else if(http.execute()) {

        }

        return true;
    }

    @Override
    public boolean Mesibo_onStartFileTransfer(MesiboFileTransfer file) {
        if(!file.upload)
            return Mesibo_onStartDownload(file);

        return Mesibo_onStartUpload(file);
    }

    @Override
    public boolean Mesibo_onStopFileTransfer(MesiboFileTransfer file) {
        MesiboHttp http = (MesiboHttp) file.getFileTransferContext();
        if(null != http)
            http.cancel();

        return true;
    }

    private int mDownloadCounter = 0, mUploadCounter = 0;
    public synchronized int updateDownloadCounter(int increment) {
        mDownloadCounter += increment;
        return mDownloadCounter;
    }

    public synchronized int updateUploadCounter(int increment) {
        mUploadCounter += increment;
        return mUploadCounter;
    }
}

File: app/src/main/java/org/mesibo/messenger/MesiboFileTransferHelper.kt
/** Copyright (c) 2019 Mesibo
 * https://mesibo.com
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the terms and condition mentioned on https://mesibo.com
 * as well as following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list
 * of conditions, the following disclaimer and links to documentation and source code
 * repository.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or other
 * materials provided with the distribution.
 *
 * Neither the name of Mesibo nor the names of its contributors may be used to endorse
 * or promote products derived from this software without specific prior written
 * permission.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
 * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Documentation
 * https://mesibo.com/documentation/
 *
 * Source Code Repository
 * https://github.com/mesibo/messengerKotlin-app-android
 *
 */

package org.mesibo.messenger

import android.os.Bundle

import com.google.gson.Gson
import com.mesibo.api.Mesibo

class MesiboFileTransferHelper internal constructor() : Mesibo.FileTransferHandler {

    private var mDownloadCounter = 0
    private var mUploadCounter = 0

    class MesiboUrl internal constructor() {
        var op: String? = null
        var file: String? = null
        var result: String? = null
        var xxx: String? = null

        init {
            result = null
            op = null
            xxx = null
            file = null
        }
    }

    init {
        Mesibo.addListener(this)
    }

    fun Mesibo_onStartUpload(params: Mesibo.MessageParams, file: Mesibo.FileInfo): Boolean {

        // we don't need to check origin the way we do in download
        if (Mesibo.getNetworkConnectivity() != Mesibo.CONNECTIVITY_WIFI && !file.userInteraction)
            return false

        // limit simultaneous upload or download
        if (mUploadCounter >= 3 && !file.userInteraction)
            return false

        val mid = file.mid

        val b = Bundle()
        b.putString("op", "upload")
        b.putString("token", SampleAPI.token)
        b.putLong("mid", mid)
        b.putInt("profile", 0)

        updateUploadCounter(1)
        val http = Mesibo.Http()

        http.url = SampleAPI.uploadUrl
        http.postBundle = b
        http.uploadFile = file.path
        http.uploadFileField = "photo"
        http.other = file
        file.fileTransferContext = http

        http.listener = Mesibo.HttpListener { config, state, percent ->
            val f = config.other as Mesibo.FileInfo

            if (100 == percent && Mesibo.Http.STATE_DOWNLOAD == state) {
                val response = config.dataString
                var mesibourl: MesiboUrl? = null
                try {
                    mesibourl = mGson.fromJson(response, MesiboUrl::class.java)
                } catch (e: Exception) {
                }

                if (null == mesibourl || null == mesibourl.file) {
                    Mesibo.updateFileTransferProgress(f, -1, Mesibo.FileInfo.STATUS_FAILED)
                    return@HttpListener false
                }

                //TBD, f.setPath if video is re-compressed
                f.url = mesibourl.file
            }

            var status = f.status
            if (100 == percent || status != Mesibo.FileInfo.STATUS_RETRYLATER) {
                status = Mesibo.FileInfo.STATUS_INPROGRESS
                if (percent < 0)
                    status = Mesibo.FileInfo.STATUS_RETRYLATER
            }

            if (percent < 100 || 100 == percent && Mesibo.Http.STATE_DOWNLOAD == state)
                Mesibo.updateFileTransferProgress(f, percent, status)

            if (100 == percent && Mesibo.Http.STATE_DOWNLOAD == state || status != Mesibo.FileInfo.STATUS_INPROGRESS)
                updateUploadCounter(-1)

            100 == percent && Mesibo.Http.STATE_DOWNLOAD == state || status != Mesibo.FileInfo.STATUS_RETRYLATER
        }

        mQueue?.queue(http) ?: if (http.execute()) {

        }

        return true

    }

    fun Mesibo_onStartDownload(params: Mesibo.MessageParams, file: Mesibo.FileInfo): Boolean {

        //TBD, check file type and size to decide automatic download
        if (!SampleAPI.mediaAutoDownload && Mesibo.getNetworkConnectivity() != Mesibo.CONNECTIVITY_WIFI && !file.userInteraction)
            return false

        // only realtime messages to be downloaded in automatic mode.
        if (Mesibo.ORIGIN_REALTIME != params.origin && !file.userInteraction)
            return false

        // limit simultaneous upload or download, 1st condition is redundant but for reference only
        if (Mesibo.getNetworkConnectivity() != Mesibo.CONNECTIVITY_WIFI && mDownloadCounter >= 3 && !file.userInteraction)
            return false

        updateDownloadCounter(1)

        val mid = file.mid

        var url = file.url
        if (!url.toLowerCase().startsWith("http://") && !url.toLowerCase().startsWith("https://")) {
            url = SampleAPI.downloadUrl + url
        }

        val http = Mesibo.Http()

        http.url = url
        http.downloadFile = file.path
        http.resume = true
        http.maxRetries = 10
        http.other = file
        file.fileTransferContext = http

        http.listener = Mesibo.HttpListener { http, state, percent ->
            val f = http.other as Mesibo.FileInfo

            var status = Mesibo.FileInfo.STATUS_INPROGRESS

            //TBD, we can simplify this now, don't need separate handling
            if (Mesibo.FileInfo.SOURCE_PROFILE == f.source) {
                if (100 == percent) {
                    Mesibo.updateFileTransferProgress(f, percent, Mesibo.FileInfo.STATUS_INPROGRESS)
                }
            } else {

                status = f.status
                if (100 == percent || status != Mesibo.FileInfo.STATUS_RETRYLATER) {
                    status = Mesibo.FileInfo.STATUS_INPROGRESS
                    if (percent < 0)
                        status = Mesibo.FileInfo.STATUS_RETRYLATER
                }

                Mesibo.updateFileTransferProgress(f, percent, status)

            }

            if (100 == percent || status != Mesibo.FileInfo.STATUS_INPROGRESS)
                updateDownloadCounter(-1)

            100 == percent || status != Mesibo.FileInfo.STATUS_RETRYLATER
        }

        mQueue?.queue(http) ?: if (http.execute()) {

        }

        return true
    }

    override fun Mesibo_onStartFileTransfer(file: Mesibo.FileInfo): Boolean {
        return if (Mesibo.FileInfo.MODE_DOWNLOAD == file.mode) Mesibo_onStartDownload(file.params, file) else Mesibo_onStartUpload(file.params, file)

    }

    override fun Mesibo_onStopFileTransfer(file: Mesibo.FileInfo): Boolean {
        val http = file.fileTransferContext as Mesibo.Http
        http?.cancel()

        return true
    }

    @Synchronized
    fun updateDownloadCounter(increment: Int): Int {
        mDownloadCounter += increment
        return mDownloadCounter
    }

    @Synchronized
    fun updateUploadCounter(increment: Int): Int {
        mUploadCounter += increment
        return mUploadCounter
    }

    companion object {

        private val mGson = Gson()
        private val mQueue = Mesibo.HttpQueue(4, 0)
    }
}


